sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiFlushAndFinalizeDoFn.java

```diff
@@ -70,6 +72,48 @@ public class StorageApiFlushAndFinalizeDoFn extends DoFn<KV<String, Operation>,
   private final Counter finalizeOperationsFailed =
       Metrics.counter(StorageApiFlushAndFinalizeDoFn.class, "finalizeOperationsFailed");
 
+  /** Custom exception to indicate that a stream is invalid due to an offset error. */
+  public static class StreamOffsetBeyondEndException extends IOException {
+    public StreamOffsetBeyondEndException(String message, Throwable cause) {
+      super(message, cause);
+    }
+  }
+
+  /**
+   * Checks if the given throwable is or is caused by an ApiException indicating that an offset is
+   * beyond the end of a BigQuery stream.
+   */
+  private boolean isOffsetBeyondEndOfStreamError(Throwable t) {
+    if (t == null) {
+      return false;
+    }
+    if (t instanceof ApiException) {
+      ApiException apiException = (ApiException) t;
+      if (apiException.getStatusCode().getCode() == Code.OUT_OF_RANGE) {
+        // Check if the cause is gRPC StatusRuntimeException for more specific message check
+        Throwable cause = apiException.getCause();
+        if (cause instanceof io.grpc.StatusRuntimeException) {
+          io.grpc.StatusRuntimeException grpcException = (io.grpc.StatusRuntimeException) cause;
+          return grpcException.getStatus().getCode() == io.grpc.Status.Code.OUT_OF_RANGE
```

Comment:
grpc library offers a method that can help simplify things here: [Status.fromThrowable()](https://github.com/grpc/grpc-java/blob/83538cdae3142be122496fbaa80440f39d715a47/api/src/main/java/io/grpc/Status.java#L396)

See how we use it here: https://github.com/apache/beam/blob/57b151a522e1807694daafba272e212a0c324011/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java#L773-L774